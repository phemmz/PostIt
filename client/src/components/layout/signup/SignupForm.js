import React, { Component } from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';
import sharedSignupValidations from '../../../../../server/shared/signupvalidations';
import TextFieldGroup from '../../common/TextFieldGroup';
/**
 * SignupForm class
 */
class SignupForm extends Component {
	constructor(props) {
		super(props);
		this.state = {
			username: '',
			email: '',
			password: '',
			passwordConfirmation: '',
			errors: {},
      invalid: false
		}
    /**
     * binds onChange, onSubmit and checkUserExists functions to this
     */
		this.onChange = this.onChange.bind(this);
		this.onSubmit = this.onSubmit.bind(this);
    this.checkUserExists = this.checkUserExists.bind(this);
	}
  /**
   * onChange() gets called when the input fields changes and the change is stored in the state
   */
	onChange(e) {
		this.setState({
			[e.target.name]: e.target.value
		});
	}
  /**
   * isValid() gets called on signup submit and sets any validation errors to the state
   */
  isValid() {
    const { errors, isValid } = sharedSignupValidations.commonValidations(this.state);
    if (!isValid) {
      this.setState({ errors });
    }
    return isValid;
  }
  /**
   * checkUserExists() gets called when a user leaves the username and email input field
   * It checks if the user already exist
   */
  checkUserExists(e) {
    const field = e.target.name;
    const val = e.target.value;
    if (val !== '') {
      this.props.isUserExists(val).then((res) => {
        let errors = this.state.errors;
        let invalid;
        /**
         * checks if isUserExists returns an object containing that user
         */
        if (res.data.user) {
          errors[field] = 'Another user exist with this ' + field;
          invalid = true;
        } else {
          errors[field] = '';
          invalid = false;
        }
        this.setState({ errors, invalid });
      })
    }
  }
  /**
   * onSubmit() gets called when the user clicks on the signup button
   */
	onSubmit(e) {
		e.preventDefault();
		if (this.isValid()) {
      /**
       * empties the errors object if there are no validation errors
       */
      this.setState({ errors: {} });
      /**
       * calls the userSignupRequest action that has been passed down to it through props
       * from SignupPage component
       */
      this.props.userSignupRequest(this.state).then(
        (res) => [
          this.context.router.push('/dashboard'),
          Materialize.toast('Signup Successful!!, Welcome', 4000, 'green')],
        (err) => this.setState({ errors: err.response.data })
      );
    }
  }

  render() {
    const { errors } = this.state;
    return (
      <div className="row">
        <div className="col s12 m6 offset-m3 signup">
          <form onSubmit={this.onSubmit}>
            <h4 className="green-text text-darken-4">Create An Account</h4>	
            <TextFieldGroup 
              error={errors.username}
              id="username"
              onChange={this.onChange}
              value={this.state.username}
              field="username"
              htmlFor="username"
              label="Username"
              checkUserExists={this.checkUserExists}
            />
            <TextFieldGroup 
              error={errors.email}
              id="email"
              onChange={this.onChange}
              value={this.state.email}
              field="email"
              htmlFor="email"
              label="Email"
              type="email"
              checkUserExists={this.checkUserExists}
            />
            <TextFieldGroup 
              error={errors.password}
              id="password"
              onChange={this.onChange}
              value={this.state.password}
              field="password"
              htmlFor="password"
              label="Password"
              type="password"
              checkUserExists={this.checkUserExists}
            />
            <TextFieldGroup 
              error={errors.passwordConfirmation}
              id="passwordConfirmation"
              onChange={this.onChange}
              value={this.state.passwordConfirmation}
              field="passwordConfirmation"
              htmlFor="passwordConfirmation"
              label="Confirm Password"
              type="password"
              checkUserExists={this.checkUserExists}
            />
            <div className="center">
              <button disabled={ this.state.isLoading || this.state.invalid } className="waves-effect waves-light btn">Sign Up</button>
            </div>	
          </form>	
        </div>
      </div>
    )	
  }
}

SignupForm.propTypes = {
	userSignupRequest: PropTypes.func.isRequired,
  isUserExists: PropTypes.func.isRequired 
}

SignupForm.contextTypes = {
  router: PropTypes.object.isRequired
}

export default SignupForm