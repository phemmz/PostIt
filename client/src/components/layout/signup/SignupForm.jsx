import React, { Component } from 'react';
import { browserHistory } from 'react-router';
import PropTypes from 'prop-types';
import GoogleLogin from 'react-google-login';
import sharedSignupValidations from '../../../../../server/shared/signupvalidations';
import TextFieldGroup from '../../common/TextFieldGroup.jsx';
/**
 * SignupForm class
 */
class SignupForm extends Component {
  /**
   * constructor
   * @param {*} props
   * @return {*} void
   */
  constructor(props) {
    super(props);
    this.state = {
      username: '',
      email: '',
      phoneNumber: '',
      password: '',
      passwordConfirmation: '',
      errors: {},
      invalid: false
    };
    /**
     * binds onChange, onSubmit and checkUserExists functions to this
     */
    this.onChange = this.onChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
    this.checkUserExists = this.checkUserExists.bind(this);
    this.responseGoogle = this.responseGoogle.bind(this);
  }
  /**
   * onChange() gets called when the input fields changes and the change is stored in the state
   * @param {*} event
   * @return {*} void
   */
  onChange(event) {
    this.setState({
      [event.target.name]: event.target.value
    });
  }
  /**
   * onSubmit() gets called when the user clicks on the signup button
   * @param {*} event
   * @return {*} void
   */
  onSubmit(event) {
    event.preventDefault();
    if (this.isValid()) {
      /**
       * empties the errors object if there are no validation errors
       */
      this.setState({ errors: {} });
      /**
       * calls the userSignupRequest action that has been passed down to it through props
       * from SignupPage component
       */
      this.props.userSignupRequest(this.state).then(
        () => {
          browserHistory.push('/dashboard');
          Materialize.toast('Signup Successful!!, Welcome', 4000, 'green');
        })
        .catch((err) => {
          this.setState({ errors: err.response.data });
        });
    }
  }
  /**
   * isValid() gets called on signup submit and sets any validation errors to the state
   * @return {*} isValid
   */
  isValid() {
    const { errors, isValid } = sharedSignupValidations.commonValidations(this.state);
    if (!isValid) {
      this.setState({ errors });
    }
    return isValid;
  }
  /**
   * checkUserExists() gets called when a user leaves the username and email input field
   * It checks if the user already exist
   * @param {*} event
   * @return {*} void
   */
  checkUserExists(event) {
    const field = event.target.name;
    const val = event.target.value.trim();
    if (val !== '') {
      this.props.isUserExists(val).then((res) => {
        const errors = this.state.errors;
        let invalid;
        /**
         * checks if isUserExists returns an object containing that user
         */
        if (res.data.user) {
          errors[field] = `Another user exist with this ${field}`;
          invalid = true;
        } else {
          errors[field] = '';
          invalid = false;
        }
        this.setState({ errors, invalid });
      });
    }
  }
  /**
   * Gets the userDetails as response from google
   * @param {*} response
   * @return {*} void
   */
  responseGoogle(response) {
    const userDetails = {
      username: response.profileObj.email.substring(0, response.profileObj.email.indexOf('@')),
      email: response.profileObj.email,
      password: response.profileObj.googleId,
      phoneNumber: response.profileObj.googleId
    };
    this.props.googleAuthentication(userDetails)
      .then(() => {
        browserHistory.push('/dashboard');
        Materialize.toast('Welcome!!', 4000, 'green');
      });
  }
/**
 * render
 * @return {*} div
 */
  render() {
    const { errors } = this.state;
    return (
      <div className="row">
        <div className="col s12 m6 offset-m3 signup">
          <div className="right">
            <GoogleLogin
              clientId="484298663558-arrp0kt4u2j2aro9k2bbb2bcffth9fke.apps.googleusercontent.com"
              onSuccess={this.responseGoogle}
              onFailure={this.responseGoogle}
              className="google-link"
              buttonText="Signup with Google"
            />
          </div>
          <form onSubmit={this.onSubmit}>
            <h4 className="green-text text-darken-4 valign">Create An Account</h4>
            <TextFieldGroup
              error={errors.username}
              id="username"
              onChange={this.onChange}
              value={this.state.username}
              field="username"
              htmlFor="username"
              label="Username"
              checkUserExists={this.checkUserExists}
            />
            <TextFieldGroup
              error={errors.email}
              id="email"
              onChange={this.onChange}
              value={this.state.email}
              field="email"
              htmlFor="email"
              label="Email"
              type="email"
              checkUserExists={this.checkUserExists}
            />
            <TextFieldGroup
              error={errors.phoneNumber}
              id="phoneNumber"
              onChange={this.onChange}
              value={this.state.phoneNumber}
              field="phoneNumber"
              htmlFor="phoneNumber"
              label="Phone Number"
              checkUserExists={this.checkUserExists}
            />
            <TextFieldGroup
              error={errors.password}
              id="password"
              onChange={this.onChange}
              value={this.state.password}
              field="password"
              htmlFor="password"
              label="Password"
              type="password"
            />
            <TextFieldGroup
              error={errors.passwordConfirmation}
              id="passwordConfirmation"
              onChange={this.onChange}
              value={this.state.passwordConfirmation}
              field="passwordConfirmation"
              htmlFor="passwordConfirmation"
              label="Confirm Password"
              type="password"
            />
            <div className="center">
              <button
                disabled={this.state.isLoading || this.state.invalid}
                className="waves-effect waves-light btn"
              >Sign Up</button>
            </div>
          </form>
        </div>
      </div>
    );
  }
}

SignupForm.propTypes = {
  userSignupRequest: PropTypes.func.isRequired,
  isUserExists: PropTypes.func.isRequired,
  googleAuthentication: PropTypes.func.isRequired
};

export default SignupForm;
