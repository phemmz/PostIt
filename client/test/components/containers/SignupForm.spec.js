import React from 'react';
import { mount } from 'enzyme';
import SignupForm from
  '../../../src/components/layout/signup/SignupForm';
import AuthenticationActions from
  '../../../src/actions/authActions';
import { initialSignupState } from '../../__mockData__/dummyData';

jest.mock('react-google-login');

const setup = () => {
  const props = {
    userSignupRequest: () => Promise.resolve(),
    isUserExists: () => Promise.resolve(),
    googleAuthentication: jest.fn(),
    onSubmit: jest.fn()
  };
  const enzymeWrapper = mount(
    <SignupForm {...props} />
  );
  return {
    props,
    enzymeWrapper
  };
};

describe('SignupForm Component', () => {
  const { props, enzymeWrapper } = setup();
  it('should render a form for signin up', () => {
    expect(enzymeWrapper).toMatchSnapshot();
    expect(enzymeWrapper.exists()).toBe(true);
    expect(enzymeWrapper.find('div').exists()).toBe(true);
    expect(enzymeWrapper.find('form').exists()).toBe(true);
    expect(enzymeWrapper.find('h4').exists()).toBe(true);
    expect(enzymeWrapper.find('TextFieldGroup')).toHaveLength(5);
  });
  it('should have an initial state', () => {
    expect(enzymeWrapper.state()).toEqual(initialSignupState);
  });
  it('should set state on change of input fields', () => {
    enzymeWrapper.find('#username').simulate('change', {
      target: { name: 'username', value: 'user1' }
    });
    expect(enzymeWrapper.state('username')).toBe('user1');
    enzymeWrapper.find('#email').simulate('change', {
      target: { name: 'email', value: 'user1@gmail.com' }
    });
    expect(enzymeWrapper.state('email')).toBe('user1@gmail.com');
    enzymeWrapper.find('#phoneNumber').simulate('change', {
      target: { name: 'phoneNumber', value: '0801234567' }
    });
    expect(enzymeWrapper.state('phoneNumber')).toBe('0801234567');
    enzymeWrapper.find('#password').simulate('change', {
      target: { name: 'password', value: '123456' }
    });
    expect(enzymeWrapper.state('password')).toBe('123456');
    enzymeWrapper.find('#passwordConfirmation').simulate('change', {
      target: { name: 'passwordConfirmation', value: '123456' }
    });
    expect(enzymeWrapper.state('passwordConfirmation')).toBe('123456');
  });
  it('should check if a user exist onBlur', () => {
    const event = { target: { name: 'username', value: 'user1' } };
    enzymeWrapper.find('#username').simulate('blur', event);
    expect(enzymeWrapper.state('invalid')).toBe(false);
  });
  it('should update the state when signup button is clicked', () => {
    const form = enzymeWrapper.find('form').first();
    form.simulate('submit', {
      preventDefault: () => {
      },
      target: [
        {
          name: 'username',
          value: 'user1',
        }
      ],
    });
    expect(enzymeWrapper.state('username')).toBe('user1');
  });
});