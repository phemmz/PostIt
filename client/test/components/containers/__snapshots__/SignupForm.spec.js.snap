// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignupForm Component should render a form for signin up 1`] = `
<SignupForm
  googleAuthentication={[Function]}
  isUserExists={[Function]}
  onSubmit={[Function]}
  userSignupRequest={[Function]}
>
  <div
    className="row"
  >
    <div
      className="col s12 m6 offset-m3 signup"
    >
      <div
        className="right"
      >
        <t
          buttonText="Signup with Google"
          className="google-link"
          clientId="484298663558-arrp0kt4u2j2aro9k2bbb2bcffth9fke.apps.googleusercontent.com"
          cookiePolicy="single_host_origin"
          disabledStyle={
            Object {
              "opacity": 0.6,
            }
          }
          fetchBasicProfile={true}
          isSignedIn={false}
          onFailure={[Function]}
          onRequest={[Function]}
          onSuccess={[Function]}
          prompt=""
          responseType="permission"
          scope="profile email"
          tag="button"
          uxMode="popup"
        />
      </div>
      <form
        onSubmit={[Function]}
      >
        <h4
          className="green-text text-darken-4 valign"
        >
          Create An Account
        </h4>
        <TextFieldGroup
          checkUserExists={[Function]}
          error=""
          field="username"
          htmlFor="username"
          id="username"
          label="Username"
          onChange={[Function]}
          type="text"
          value=""
        >
          <div
            className="input-field col s12"
          >
            <input
              className="form-control"
              id="username"
              name="username"
              onBlur={[Function]}
              onChange={[Function]}
              required={true}
              type="text"
              value=""
            />
            <label
              htmlFor="username"
            >
              Username
            </label>
          </div>
        </TextFieldGroup>
        <TextFieldGroup
          checkUserExists={[Function]}
          error=""
          field="email"
          htmlFor="email"
          id="email"
          label="Email"
          onChange={[Function]}
          type="email"
          value=""
        >
          <div
            className="input-field col s12"
          >
            <input
              className="form-control"
              id="email"
              name="email"
              onBlur={[Function]}
              onChange={[Function]}
              required={true}
              type="email"
              value=""
            />
            <label
              htmlFor="email"
            >
              Email
            </label>
          </div>
        </TextFieldGroup>
        <TextFieldGroup
          checkUserExists={[Function]}
          error=""
          field="phoneNumber"
          htmlFor="phoneNumber"
          id="phoneNumber"
          label="Phone Number"
          onChange={[Function]}
          type="text"
          value=""
        >
          <div
            className="input-field col s12"
          >
            <input
              className="form-control"
              id="phoneNumber"
              name="phoneNumber"
              onBlur={[Function]}
              onChange={[Function]}
              required={true}
              type="text"
              value=""
            />
            <label
              htmlFor="phoneNumber"
            >
              Phone Number
            </label>
          </div>
        </TextFieldGroup>
        <TextFieldGroup
          checkUserExists={[Function]}
          error=""
          field="password"
          htmlFor="password"
          id="password"
          label="Password"
          onChange={[Function]}
          type="password"
          value=""
        >
          <div
            className="input-field col s12"
          >
            <input
              className="form-control"
              id="password"
              name="password"
              onBlur={[Function]}
              onChange={[Function]}
              required={true}
              type="password"
              value=""
            />
            <label
              htmlFor="password"
            >
              Password
            </label>
          </div>
        </TextFieldGroup>
        <TextFieldGroup
          checkUserExists={[Function]}
          error=""
          field="passwordConfirmation"
          htmlFor="passwordConfirmation"
          id="passwordConfirmation"
          label="Confirm Password"
          onChange={[Function]}
          type="password"
          value=""
        >
          <div
            className="input-field col s12"
          >
            <input
              className="form-control"
              id="passwordConfirmation"
              name="passwordConfirmation"
              onBlur={[Function]}
              onChange={[Function]}
              required={true}
              type="password"
              value=""
            />
            <label
              htmlFor="passwordConfirmation"
            >
              Confirm Password
            </label>
          </div>
        </TextFieldGroup>
        <div
          className="center"
        >
          <button
            className="waves-effect waves-light btn"
            disabled={false}
          >
            Sign Up
          </button>
        </div>
      </form>
    </div>
  </div>
</SignupForm>
`;
